### PR TITLE
MINIMUM_STEPPER_PULSE to MINIMUM_STEPPER_PULSE_NS

### PR DESCRIPTION
**Pull request by mh-dm**
### Description

Switch from `MINIMUM_STEPPER_PULSE` (in µs) to `MINIMUM_STEPPER_PULSE_NS` (in nanoseconds). This PR is basically just code refactoring. Note that `MINIMUM_STEPPER_PULSE` is in `Configuration_adv.h` and if left undefined it is automatically configured by `Conditionals_adv.h` based on the used stepper drivers.

### Benefits

Cleaner code for handling stepper motor drivers that can do sub microsecond step pulse widths (like LV8729 and TMC2xxx). Allows easier user experimentation with sub microsecond step pulse widths.

### CODE REVIEW

**Comment by thisiskeithb** [comment]
Since a config option has changed / been renamed, you’ll need to add it to [`Changes.h`](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/inc/Changes.h).

**Comment by mh-dm** [comment]
Thanks, done.

File: Marlin/src/module/stepper.cpp [hunk 1]
```diff
@@ -529,10 +529,10 @@ xyze_int8_t Stepper::count_direction{0};
 // Round up when converting from ns to timer ticks
 #define NS_TO_PULSE_TIMER_TICKS(NS) (((NS) + (NS_PER_PULSE_TIMER_TICK) / 2) / (NS_PER_PULSE_TIMER_TICK))
 
-#define TIMER_SETUP_NS (CYCLES_TO_NS(TIMER_READ_ADD_AND_STORE_CYCLES))
+#define TIMER_SETUP_NS CYCLES_TO_NS(timer_read_add_and_store_cycles)
 
-#define PULSE_HIGH_TICK_COUNT hal_timer_t(NS_TO_PULSE_TIMER_TICKS(_MIN_PULSE_HIGH_NS - _MIN(_MIN_PULSE_HIGH_NS, TIMER_SETUP_NS)))
-#define PULSE_LOW_TICK_COUNT hal_timer_t(NS_TO_PULSE_TIMER_TICKS(_MIN_PULSE_LOW_NS - _MIN(_MIN_PULSE_LOW_NS, TIMER_SETUP_NS)))
+#define PULSE_HIGH_TICK_COUNT hal_timer_t(NS_TO_PULSE_TIMER_TICKS(_min_pulse_high_ns - _MIN(_min_pulse_high_ns, TIMER_SETUP_NS)))
+#define PULSE_LOW_TICK_COUNT hal_timer_t(NS_TO_PULSE_TIMER_TICKS(_min_pulse_low_ns - _MIN(_min_pulse_low_ns, TIMER_SETUP_NS)))
```
Comment (mh-dm) [inline Marlin/src/module/stepper.cpp]: If we're doing constexpr then `constexpr hal_timer_t PULSE_HIGH_TICK_COUNT = [..];` and similar gets rid of the cast.

File: Marlin/src/module/stepper/cycles.h [hunk 2]
```diff
@@ -40,177 +40,127 @@
...
+                   isr_u_stepper_cycles = TERN0(HAS_U_STEP, isr_stepper_cycles),
+                   isr_v_stepper_cycles = TERN0(HAS_V_STEP, isr_stepper_cycles),
+                   isr_w_stepper_cycles = TERN0(HAS_W_STEP, isr_stepper_cycles),
+                   isr_e_stepper_cycles = TERN0(HAS_EXTRUDERS, isr_stepper_cycles); // E is always interpolated, even for mixing extruders
 
 // And the total minimum loop time, not including the base
-#define _PLUS_AXIS_CYCLES(A) + (ISR_##A##_STEPPER_CYCLES)
-#define MIN_ISR_LOOP_CYCLES (ISR_MIXING_STEPPER_CYCLES LOGICAL_AXIS_MAP(_PLUS_AXIS_CYCLES))
+#define _PLUS_AXIS_CYCLES(a) + (isr_##a##_stepper_cycles)
+constexpr uint32_t min_isr_loop_cycles = (isr_mixing_stepper_cycles LOGICAL_AXIS_MAP_LC(_PLUS_AXIS_CYCLES));
```
Comment (mh-dm) [inline Marlin/src/module/stepper/cycles.h]: I was about to comment that the macro will break with lower case before I noticed the `_LC`. I think it's worth considering keeping SCREAMING_CASE when switching from `#define` to `constexpr` when macros are involved, at least as a visual cue that macros are (still) involved. (There's also the git blame/code history and merge conflicts arguments but you've probably already considered them)

Comment (thinkyhead) [reply Marlin/src/module/stepper/cycles.h]: Lowercase for things that aren't a `#define` makes it easy to distinguish that they are not macros and can be treated/trusted to be typed values, so that is generally preferred. We may still uppercase that axis letter, but also I want to change those individual defined values into an array indexed by axis, so stay tuned….

**Comment by thinkyhead** [comment]
Some of these things defined as `hal_timer_t` should probably be changed to `uint32_t`. Both AVR and HC32 use 16-bit timer width, so it's very likely to overflow. If we're curious, we could add tests first for to see if any `hal_timer_t` value is currently overflowing `HAL_TIMER_TYPE_MAX` and add a build test with plausible high step time values to see if that leads to overflow.

File: Marlin/src/HAL/TEENSY40_41/timers.h [hunk 3]
```diff
@@ -36,13 +36,13 @@
 typedef uint32_t hal_timer_t;
 #define HAL_TIMER_TYPE_MAX 0xFFFFFFFE
 
-#define GPT_TIMER_RATE F_BUS_ACTUAL   // 150MHz
+#define GPT_TIMER_RATE F_BUS // 150MHz (Can't use F_BUS_ACTUAL because it's extern volatile)
```
Comment (thinkyhead) [inline Marlin/src/HAL/TEENSY40_41/timers.h]: Note that this is only 1/4 of the full `F_CPU` speed, but it's unclear whether we can run interrupt timers at such a high resolution as 1/1, or that we need to. All counts would require two extra bits.

Since `F_BUS_ACTUAL` is `volatile`, perhaps the system clock speed is able to change dynamically, or maybe this is only done under program control. If Marlin is able to keep the clock at a constant rate then we should be ok just using `F_BUS`, which is provided by the Teensy platform. I'm not sure yet how `F_CPU` and `F_BUS` are defined, since you can change the CPU speed with PlatformIO option `board_build.f_cpu`. If Teensy isn't picking up `board_build.f_cpu` then we may add our own `-DF_CPU` or some other define that Teensy platform is expecting.

File: Marlin/src/module/stepper/cycles.h [hunk 4]
```diff
@@ -126,20 +126,20 @@ constexpr uint32_t min_isr_loop_cycles = (isr_mixing_stepper_cycles LOGICAL_AXIS
 #endif
 
 // Calculate the minimum MPU cycles needed per pulse to enforce
-constexpr uint32_t min_stepper_pulse_cycles = _min_pulse_high_ns * CYCLES_PER_MICROSECOND / 1000;
+constexpr hal_timer_t min_stepper_pulse_cycles = _min_pulse_high_ns * CYCLES_PER_MICROSECOND / 1000;
```
Comment (mh-dm) [inline Marlin/src/module/stepper/cycles.h]: This commit is problematic. `MIN_STEPPER_PULSE_CYCLES` is definitely `uint32_t` (cycles) and not `hal_timer_t` i.e. timer ticks (usually cycles/4).
I think basically all of the changed lines in this file are cycles and not `hal_timer_t`.

File: Marlin/src/module/stepper/cycles.h [hunk 5]
```diff
@@ -133,7 +133,7 @@ constexpr uint32_t min_stepper_pulse_cycles = _min_pulse_high_ns * CYCLES_PER_MI
 constexpr uint32_t isr_loop_cycles(const int R) { return ((isr_loop_base_cycles + min_isr_loop_cycles + min_stepper_pulse_cycles) * ((1UL << R) - 1) + _MAX(min_isr_loop_cycles, min_stepper_pulse_cycles)); }
 
 // Model input shaping as an extra loop call
-constexpr uint32_t isr_shaping_loop_cycles(const int R) { return (TERN0(HAS_ZV_SHAPING, (isr_loop_base_cycles + TERN0(INPUT_SHAPING_X, isr_x_stepper_cycles) + TERN0(INPUT_SHAPING_Y, isr_y_stepper_cycles)) << R)); }
+constexpr uint32_t isr_shaping_loop_cycles(const int R) { return (TERN0(HAS_ZV_SHAPING, (isr_loop_base_cycles + TERN0(INPUT_SHAPING_X, isr_x_stepper_cycles) + TERN0(INPUT_SHAPING_Y, isr_y_stepper_cycles) + TERN0(INPUT_SHAPING_Y, isr_z_stepper_cycles)) << R)); }
```
Comment (mh-dm) [inline Marlin/src/module/stepper/cycles.h]: `INPUT_SHAPING_Z`

**Comment by mh-dm** [comment]
> Some of these things defined as `hal_timer_t` should probably be changed to `uint32_t`. Both AVR and HC32 use 16-bit timer width, so it's very likely to overflow. If we're curious, we could add tests first for to see if any `hal_timer_t` value is currently overflowing `HAL_TIMER_TYPE_MAX` and add a build test with plausible high step time values to see if that leads to overflow.

`hal_timer_t` is for timer ticks which is usually `cycles/4`. And given that most things changed look like cycles they should indeed be `uint32_t` instead.

If there is indeed some math / comparison being done between timer ticks/`hal_timer_t` and cycles/`uint32_t` without a clear conversion being done then that's revealing of a bug. If my PR didn't introduce it I recommend it be addressed in a separate PR since this one is already getting too big for my comfort :).

File: Marlin/src/module/stepper/cycles.h (hunk 4 above)
Comment (thinkyhead) [reply Marlin/src/module/stepper/cycles.h]: Please point out any others that should be `uint32_t` which were missed. As I mentioned in another comment, we will surely need 32-bits for any timer values that are being multiplied or shifted so they don't overflow on AVR. We can also use a minimum size, which would make `hal_timer_t` 32-bits on all platforms where it isn't 64 bits.

File: Marlin/src/module/stepper.cpp [hunk 6]
```diff
@@ -519,16 +519,16 @@ xyze_int8_t Stepper::count_direction{0};
   #define E_APPLY_DIR(FWD,Q) do{ if (FWD) { FWD_E_DIR(stepper_extruder); } else { REV_E_DIR(stepper_extruder); } }while(0)
 #endif
 
-#define CYCLES_TO_NS(CYC) (1000UL * (CYC) / ((F_CPU) / 1000000))
-#define NS_PER_PULSE_TIMER_TICK (1000000000UL / (STEPPER_TIMER_RATE))
+constexpr uint32_t cycles_to_ns(const hal_timer_t CYC) { return 1000UL * (CYC) / ((F_CPU) / 1000000); }
```
Comment (mh-dm) [inline Marlin/src/module/stepper.cpp]: Cycles so `uint32_t CYC`

File: Marlin/src/module/stepper.cpp [hunk 7]
```diff
@@ -519,16 +519,16 @@ xyze_int8_t Stepper::count_direction{0};
...
+constexpr hal_timer_t ns_per_pulse_timer_tick = 1000000000UL / (STEPPER_TIMER_RATE);
 
 // Round up when converting from ns to timer ticks
-#define NS_TO_PULSE_TIMER_TICKS(NS) (((NS) + (NS_PER_PULSE_TIMER_TICK) / 2) / (NS_PER_PULSE_TIMER_TICK))
+constexpr hal_timer_t ns_to_pulse_timer_ticks(const uint32_t ns) { return (ns + ns_per_pulse_timer_tick / 2) / ns_per_pulse_timer_tick; }
 
-#define TIMER_SETUP_NS (CYCLES_TO_NS(TIMER_READ_ADD_AND_STORE_CYCLES))
+constexpr uint32_t timer_setup_ns = cycles_to_ns(timer_read_add_and_store_cycles);
```
Comment (mh-dm) [inline Marlin/src/module/stepper.cpp:528]: ns so `uint32_t`

File: Marlin/src/module/stepper/cycles.h [hunk 8]
```diff
@@ -40,184 +40,123 @@
...
+  #error "At least one of MINIMUM_STEPPER_PULSE_NS or MAXIMUM_STEPPER_RATE must be defined"
 #endif
 
+// Calculate the minimum MPU cycles needed per pulse to enforce
+constexpr uint32_t min_stepper_pulse_cycles = _min_pulse_high_ns * CYCLES_PER_MICROSECOND / 1000;
+
 // The loop takes the base time plus the time for all the bresenham logic for 1 << R pulses plus the time
 // between pulses for ((1 << R) - 1) pulses. But the user could be enforcing a minimum time so the loop time is:
-#define ISR_LOOP_CYCLES(R) ((ISR_LOOP_BASE_CYCLES + MIN_ISR_LOOP_CYCLES + MIN_STEPPER_PULSE_CYCLES) * ((1UL << R) - 1) + _MAX(MIN_ISR_LOOP_CYCLES, MIN_STEPPER_PULSE_CYCLES))
+constexpr hal_timer_t isr_loop_cycles(const int R) { return ((isr_loop_base_cycles + min_isr_loop_cycles + min_stepper_pulse_cycles) * ((1UL << R) - 1) + _MAX(min_isr_loop_cycles, min_stepper_pulse_cycles)); }
```
Comment (mh-dm) [inline Marlin/src/module/stepper/cycles.h]: `min_stepper_pulse_cycles` is cycles and it's just added to other `_cycles` basically confirming that everything that's called `*_cycles` in this file is `uint32_t` and not `hal_timer_t`. There are almost 20 such `_cycles` variables in this file.
Also 2 functions `isr_execution_cycles` and `isr_shaping_loop_cycles`.

File: Marlin/src/module/stepper/cycles.h [hunk 9]
```diff
@@ -40,184 +40,123 @@
...
 
 // The maximum allowable stepping frequency when doing 1x stepping (in Hz)
-#define MAX_STEP_ISR_FREQUENCY_1X ((F_CPU) / ISR_EXECUTION_CYCLES(0))
+constexpr hal_timer_t max_step_isr_frequency_1x = (F_CPU) / isr_execution_cycles(0);
+constexpr hal_timer_t max_step_isr_frequency_sh(const int S) { return ((F_CPU) / isr_execution_cycles(S)) >> S; }
 
 // The minimum step ISR rate used by ADAPTIVE_STEP_SMOOTHING to target 50% CPU usage
 // This does not account for the possibility of multi-stepping.
-#define MIN_STEP_ISR_FREQUENCY (MAX_STEP_ISR_FREQUENCY_1X >> 1)
+constexpr hal_timer_t min_step_isr_frequency = max_step_isr_frequency_1x >> 1;
```
Comment (mh-dm) [inline Marlin/src/module/stepper/cycles.h]: These are frequencies therefore `uint32_t`.

File: Marlin/src/module/stepper.cpp [hunk 10]
```diff
@@ -519,16 +519,16 @@ xyze_int8_t Stepper::count_direction{0};
   #define E_APPLY_DIR(FWD,Q) do{ if (FWD) { FWD_E_DIR(stepper_extruder); } else { REV_E_DIR(stepper_extruder); } }while(0)
 #endif
 
-#define CYCLES_TO_NS(CYC) (1000UL * (CYC) / ((F_CPU) / 1000000))
-#define NS_PER_PULSE_TIMER_TICK (1000000000UL / (STEPPER_TIMER_RATE))
+constexpr uint32_t cycles_to_ns(const hal_timer_t CYC) { return 1000UL * (CYC) / ((F_CPU) / 1000000); }
+constexpr hal_timer_t ns_per_pulse_timer_tick = 1000000000UL / (STEPPER_TIMER_RATE);
```
Comment (mh-dm) [inline Marlin/src/module/stepper.cpp]: ns so `uint32_t`

**Review by mh-dm — COMMENTED** [review]
I did a pass over everything at 92fe629 commit and it looks good.

File: Marlin/src/core/types.h [hunk 11]
```diff
@@ -274,6 +274,13 @@ enum AxisEnum : uint8_t {
 #define LOOP_DISTINCT_AXES(VAR) for (uint8_t VAR = 0; VAR < DISTINCT_AXES; ++VAR)
 #define LOOP_DISTINCT_E(VAR) for (uint8_t VAR = 0; VAR < DISTINCT_E; ++VAR)
 
+// Time, ticks, cycles
+typedef uint32_t cycles_t;  // Cycle counting may be larger than timer size
+typedef uint32_t ticks_t;   // Tick counting, according to each timer's rate
+typedef uint32_t micros_t;  // Microseconds
+typedef uint32_t nanos_t;   // Nanoseconds
+typedef uint32_t hertz_t;   // Hertz
```
Comment (mh-dm) [inline Marlin/src/core/types.h]: To expand on why I personally don't like these: visual clutter, slightly harder to process mentally since they're unfamiliar (and will be to anyone contributing), duplication of information as we already have _ns and _us endings. Particular to `ticks_t`, now there's multiple tick types counting at different rates. The compiler doesn't help us by giving warnings when we directly add/subtract differing units together. Now the code can be wrong / inconsistent in more ways (using uint32_t when should have used micros_t for example). The size and being unsigned is slightly hidden at point of use which makes it easier to overlook potentially problematic automatic int->uint conversions.

**Comment by thinkyhead** [comment]
> I did a pass over everything…

Thanks for the review. If more variables, methods, and parameters should be working with `uint32_t` (especially under known circumstances) we can certainly apply that. For AVR we generally want to use the smallest types required, so that is specifically where `hal_timer_t` comes in. It generally makes sense to convert `#define` values to 32-bit and let the compiler decide how to deal with type conversion. Any of the `constexpr uint32_t` defined with a value under 65536 could be changed to `uint16_t` and that may be better for some code on AVR, or it may make no difference. As we are further fixing up and optimizing the motion code with these PRs I won't sweat that small detail until some profiling shows that it matters.

**Comment by thisiskeithb** [comment]
This broke all our [Artillery configs](https://github.com/MarlinFirmware/Configurations/tree/bugfix-2.1.x/config/examples/Artillery) with TMC2100s:

- https://github.com/MarlinFirmware/Configurations/issues/1068

I think I found the fix:

These configs use TMC2100s and I cannot find a `HAS_DRIVER(TMC2100)` check anywhere within Marlin, so `HAS_TRINAMIC_CONFIG` or `HAS_TRINAMIC_STANDALONE` is never set, thus no `MINIMUM_STEPPER_PULSE_NS` is set in `Conditionals_adv.h`.

Semi-related, looking at the TMC2100 option some more, we only have a single driver type option (`TMC2100`), but it is really `TMC2100_STANDALONE`:

image

---

PR submitted: https://github.com/MarlinFirmware/Marlin/pull/27209
